Feature Add Monotonic Definition

mertak-synnada opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #.

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->
This PR creates a definition of set-monotonicity for Aggregate expressions. Some aggregation functions create ordered results by definition (such as count, min, max). With this PR, we're adding this information to the output ordering and be able to remove some SortExecs while optimizing

## What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

## Are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

## Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

Commit:
add monotonic function definitions for aggregate expressions

mertak-synnada commented on datafusion/physical-expr/src/aggregate.rs:
```diff
@@ -533,6 +536,29 @@ impl AggregateFunctionExpr {
+        let idx = schema.index_of(self.name()).unwrap_or(0);
+        let expr = Arc::new(Column::new(self.name(), idx));
+
+        let options = SortOptions::new(!is_ascending, false);
```
We assumed Nulls Last by definition, I'm not fully sure about this will be as is in the future as well

mertak-synnada commented on datafusion/expr/src/udaf.rs:
```diff
@@ -635,6 +635,18 @@ pub trait AggregateUDFImpl: Debug + Send + Sync {
+    /// Returns None if the function is not monotonic.
+    /// If the function is monotonically decreasing returns Some(false) e.g. Min
+    /// If the function is monotonically increasing returns Some(true) e.g. Max
+    fn is_monotonic(&self) -> Option<bool> {
```
I think we will need to accept a `data_type` argument here since `Sum`s monotonicity depends on it.

jayzhan-synnada replied:
we can add it first so we don't have to break API again in the future

mertak-synnada commented on datafusion/physical-expr/src/aggregate.rs:
```diff
@@ -533,6 +536,29 @@ impl AggregateFunctionExpr {
+    pub fn natural_sort_expr(&self, schema: &SchemaRef) -> Option<PhysicalSortExpr> {
+        // If the aggregate expressions are monotonic, the output data is naturally ordered with it.
+        let is_ascending = self.is_monotonic()?;
+        let idx = schema.index_of(self.name()).unwrap_or(0);
```
I don't like the way we get the `expr` here but couldn't find a better way

jayzhan-synnada replied:
return Result for this function and not return incorrect 0 index seems better

mertak-synnada replied:
I tried to implement so but got too many errors on tests, the expression name does not always match the actual column name. But there should not be any need for us to search with name, I'll try to get the actual index to here and insert

mertak-synnada commented on datafusion/sqllogictest/test_files/aggregate.slt:
```diff
@@ -4978,16 +4978,17 @@ logical_plan
-11)--------------------RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=1
-12)----------------------CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/testing/data/csv/aggregate_test_100.csv]]}, projection=[c1, c3], has_header=true
+03)----AggregateExec: mode=FinalPartitioned, gby=[c3@0 as c3, min(aggregate_test_100.c1)@1 as min(aggregate_test_100.c1)], aggr=[], ordering_mode=PartiallySorted([1])
+04)------SortExec: expr=[min(aggregate_test_100.c1)@1 ASC NULLS LAST], preserve_partitioning=[true]
```
This SortExec is being added, because there's another logic on AggregateExec's partition_by values. The parent `AggregateExec` is adding the min column as an ordering requirement, and since `prefer_existing_sort` is false by default `RepartitionExec` does not maintain input order.

Commit:
fix benchmark results

Commit:
set prefer_existing_sort to true in sqllogictests

Commit:
set prefer_existing_sort to true in sqllogictests

Commit:
fix typo

Commit:
Merge branch 'refs/heads/apache_main' into feature/monotonic-sets

# Conflicts:
#	datafusion/core/src/physical_optimizer/enforce_sorting.rs
#	datafusion/core/src/physical_optimizer/test_utils.rs

Commit:
re-add test_utils.rs changes to the new file

Commit:
clone input with Arc

Commit:
Merge branch 'refs/heads/apache_main' into feature/monotonic-sets

alihan-synnada commented:
Small nit: `is_monotonic` returning `Option<bool>` is a little confusing. The documentation is clear about what the returned value means so it isn't a huge deal. I don't want to waste too much time bikeshedding but here are a few alternatives:

- `is_monotonically_increasing()` sounds more descriptive but it's rather long
-  `monotonicity()` could return an enum with 3 variants Monotonicity(NonDecreasing, NonIncreasing, NonMonotonic)

LGTM otherwise 👍

jayzhan-synnada commented on datafusion/physical-optimizer/src/test_utils.rs:
```diff
@@ -207,6 +251,56 @@ pub fn aggregate_exec(input: Arc<dyn ExecutionPlan>) -> Arc<dyn ExecutionPlan> {
+        ];
+    Arc::new(
+        AggregateExec::try_new(
+            AggregateMode::Final,
```
```suggestion
            AggregateMode::Single,
```

Final is used in 2-step aggregation, should be Single here

jayzhan-synnada commented on datafusion/core/src/physical_optimizer/enforce_sorting.rs:
```diff
@@ -1010,6 +1023,63 @@ mod tests {
+        let expected_input = [
+            "SortExec: expr=[avg@0 ASC], preserve_partitioning=[false]",
+            "  AggregateExec: mode=Final, gby=[], aggr=[avg]",
+            "    SortExec: expr=[a@0 ASC], preserve_partitioning=[false]",
```
I guess `avg` is monotonic for integer but not for float?

mertak-synnada replied:
Nope, it's not monotonic for unsigned integers, for example:

1st data 100 => Avg 100
2nd data 2 => (102/2) Avg 51 (decreased)
3rd data 3 => (105/3) Avg 35 (decreased)
4th data 1003 => (1108/4) Avg 277 (increased)

To be monotonic, it needs to be always increasing/decreasing

jayzhan-synnada commented on datafusion/physical-optimizer/src/test_utils.rs:
```diff
@@ -207,6 +251,56 @@ pub fn aggregate_exec(input: Arc<dyn ExecutionPlan>) -> Arc<dyn ExecutionPlan> {
+        ];
+    Arc::new(
+        AggregateExec::try_new(
+            AggregateMode::Final,
```
```suggestion
            AggregateMode::Single,
```

jayzhan-synnada commented on datafusion/physical-expr/src/window/aggregate.rs:
```diff
@@ -67,6 +67,16 @@ impl PlainAggregateWindowExpr {
+    pub fn add_equal_orderings(&self, eq_properties: &mut EquivalenceProperties) {
+        let Some(expr) = self
+            .get_aggregate_expr()
+            .natural_sort_expr(eq_properties.schema())
```
Does anyone know why `schema` is inside `EquivalenceProperties` but not `PlanProperties` 🤔  or even not storing them inside both of them but pass as a parameter instead.

mertak-synnada replied:
🤷‍♂️ Have no idea, but I'll remove the schema need anyways

Commit:
Merge branch 'refs/heads/apache_main' into feature/monotonic-sets

# Conflicts:
#	datafusion/core/src/physical_optimizer/enforce_sorting.rs
#	datafusion/physical-optimizer/src/test_utils.rs

Commit:
inject aggr expr indices
separate stubs and count_udafs

Commit:
remove redundant file

Commit:
add Sum monotonicity
change monotonicity to return an Enum rather than Option<bool>
fix indices
re-add monotonicity tests

Commit:
fix sql logic tests

Commit:
fix sql logic tests

Commit:
Merge branch 'refs/heads/apache_main' into feature/monotonic-sets

# Conflicts:
#	datafusion/core/tests/physical_optimizer/enforce_sorting.rs

Commit:
update docs

berkaysynnada committed:
Merge branch 'apache_main' into feature/monotonic-sets

berkaysynnada committed:
review part 1

buraksenn committed:
make AnalysisContext aware of empty sets to represent certainly false bounds (#14279)

* ready for review

* fmt and lint

* Apply suggestions from code review

Co-authored-by: Mehmet Ozan Kabak <ozankabak@gmail.com>

* apply reviews

* fix test

* Update analysis.rs

* Update analysis.rs

---------

Co-authored-by: Mehmet Ozan Kabak <ozankabak@gmail.com>
Co-authored-by: berkaysynnada <berkay.sahin@synnada.ai>

phisn committed:
Add relation to alias expr in schema display (#14311)

* Add relation to alias expr in schema display

* Add edge case for alias in physical_name

alamb committed:
Improve deprecation message for MemoryExec (#14322)

* Improve deprecation message for `MemoryExec`

* clippy

dhegberg committed:
fix: LogicalPlan::get_parameter_types fails to return all placeholders (#14312)

zhuqi-lucas committed:
fix: FULL OUTER JOIN and LIMIT produces wrong results (#14338)

* fix: FULL OUTER JOIN and LIMIT produces wrong results

* Fix minor slt testing

* fix test

Sevenannn committed:
Customize window frame support for dialect (#14288)

* Customize window frame support for dialect (#70)

* Customize window frame support for dialect

* fix: ignore frame only when frame implies no frame

* Add comments. move the window frame determine logic to dialect method

* Update datafusion/sql/src/unparser/dialect.rs

* Update test case

* fix

---------

Co-authored-by: Phillip LeBlanc <phillip@leblanc.tech>

* fix clippy

---------

Co-authored-by: Phillip LeBlanc <phillip@leblanc.tech>

Chen-Yuan-Lai committed:
refactor: switch BooleanBufferBuilder to NullBufferBuilder in a unit test of common_scalar (#14339)

Co-authored-by: Cheng-Yuan-Lai <a186235@g,ail.com>

gatesn committed:
Add `ColumnStatistics::Sum` (#14074)

* Add sum statistic

* Add sum statistic

* Add sum statistic

* Add sum statistic

* Add sum statistic

* Add sum statistic

* Add tests and Cargo fmt

* fix up

---------

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

Chen-Yuan-Lai committed:
refactor: switch BooleanBufferBuilder to NullBufferBuilder in unit tests for unnest (#14321)

Co-authored-by: Cheng-Yuan-Lai <a186235@g,ail.com>

alamb committed:
Increase MSRV to 1.81.0 (#14330)

* Increase MSRV to 1.81.0

* also update datafusion-cli

alamb committed:
Fix build due to logical error (#14345)

berkaysynnada committed:
fix the tests

berkaysynnada committed:
revert slt's

berkaysynnada committed:
simplify terms

berkaysynnada committed:
Update mod.rs

berkaysynnada committed:
remove unnecessary computations

berkaysynnada committed:
remove index calc

berkaysynnada committed:
Update mod.rs

ozankabak committed:
Apply suggestions from code review

berkaysynnada committed:
add slt

dependabot[bot] committed:
chore(deps): bump serde_json from 1.0.137 to 1.0.138 in /datafusion-cli (#14351)

Bumps [serde_json](https://github.com/serde-rs/json) from 1.0.137 to 1.0.138.
- [Release notes](https://github.com/serde-rs/json/releases)
- [Commits](https://github.com/serde-rs/json/compare/v1.0.137...v1.0.138)

---
updated-dependencies:
- dependency-name: serde_json
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): bump tempfile from 3.15.0 to 3.16.0 in /datafusion-cli (#14350)

Bumps [tempfile](https://github.com/Stebalien/tempfile) from 3.15.0 to 3.16.0.
- [Changelog](https://github.com/Stebalien/tempfile/blob/master/CHANGELOG.md)
- [Commits](https://github.com/Stebalien/tempfile/compare/v3.15.0...v3.16.0)

---
updated-dependencies:
- dependency-name: tempfile
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

alamb committed:
Update version in `datafusion-cli/Dockerfile` to 1.81 (#14344)

* Update version in dockerfile

* Update datafusion-cli/Dockerfile

Co-authored-by: Oleks V <comphead@users.noreply.github.com>

---------

Co-authored-by: Oleks V <comphead@users.noreply.github.com>

rluvaton committed:
perf(array-agg): add fast path for array agg for `merge_batch` (#14299)

* perf(array-agg): add fast path for array agg for `merge_batch`

* update comment

* fix slice length

* fix: make sure we are not inserting empty lists

logan-keede committed:
moving memory.rs out of datafusion/core (#14332)

* moving memory.rs out of datafusion/core

* fix: forgotten license

* fix: taplo + documentation

* fix: cargo fmt

* Cargo.toml update, removing unnecessary dependency

Chen-Yuan-Lai committed:
refactor: switch `BooleanBufferBuilder` to `NullBufferBuilder` in binary_map (#14341)

* refactor: switch BooleanBufferBuilder to NullBufferBuilder in binary_map

* fix: check cargo.lock for datafusion-cli

* add comment for .unwrap

---------

Co-authored-by: Cheng-Yuan-Lai <a186235@g,ail.com>

findepi committed:
Restore ability to run single SLT file (#14355)

Make it possible again to run single SLT file, even if it's name is a
substring of other file(s).

For example, after the change, this command:

   cargo test --test sqllogictests -- test_files/union.slt

runs `union.slt` file, but does not run `pg_compat_union.slt`.

dependabot[bot] committed:
chore(deps): bump home from 0.5.9 to 0.5.11 in /datafusion-cli (#14257)

Bumps [home](https://github.com/rust-lang/cargo) from 0.5.9 to 0.5.11.
- [Changelog](https://github.com/rust-lang/cargo/blob/master/CHANGELOG.md)
- [Commits](https://github.com/rust-lang/cargo/compare/home-0.5.9...home-0.5.11)

---
updated-dependencies:
- dependency-name: home
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): bump aws-sdk-ssooidc in /datafusion-cli (#14314)

Bumps [aws-sdk-ssooidc](https://github.com/awslabs/aws-sdk-rust) from 1.51.0 to 1.57.1.
- [Release notes](https://github.com/awslabs/aws-sdk-rust/releases)
- [Commits](https://github.com/awslabs/aws-sdk-rust/commits)

---
updated-dependencies:
- dependency-name: aws-sdk-ssooidc
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

zhuqi-lucas committed:
fix: LimitPushdown rule uncorrect remove some GlobalLimitExec (#14245)

* fix: LimitPushdown rule uncorrect remove some GlobalLimitExec

* Fix some logic for maybe fetch

* Fix test

* Address comments

* Address comments

* Add comments

* Address comments

Chen-Yuan-Lai committed:
refactor: switch `BooleanBufferBuilder` to `NullBufferBuilder` in single_group_by (#14360)

* refactor: switch BooleanBufferBuilder to NullBufferBuilder in single_group_by

* Update datafusion/physical-plan/src/aggregates/group_values/single_group_by/primitive.rs

---------

Co-authored-by: Cheng-Yuan-Lai <a186235@g,ail.com>
Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

erenavsarogullari committed:
Issue-14216 - Support arrays_overlap function (#14217)

dependabot[bot] committed:
chore(deps): bump rustyline from 14.0.0 to 15.0.0 in /datafusion-cli (#14265)

* chore(deps): bump rustyline from 14.0.0 to 15.0.0 in /datafusion-cli

Bumps [rustyline](https://github.com/kkawakam/rustyline) from 14.0.0 to 15.0.0.
- [Release notes](https://github.com/kkawakam/rustyline/releases)
- [Changelog](https://github.com/kkawakam/rustyline/blob/master/History.md)
- [Commits](https://github.com/kkawakam/rustyline/compare/v14.0.0...v15.0.0)

---
updated-dependencies:
- dependency-name: rustyline
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

* fix build

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: jonahgao <jonahgao@msn.com>
Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

berkaysynnada committed:
remove aggregate changes, tests already give expected results

berkaysynnada committed:
fix clippy

dependabot[bot] committed:
chore(deps): bump aws-sdk-sts from 1.51.0 to 1.57.0 in /datafusion-cli (#14263)

Bumps [aws-sdk-sts](https://github.com/awslabs/aws-sdk-rust) from 1.51.0 to 1.57.0.
- [Release notes](https://github.com/awslabs/aws-sdk-rust/releases)
- [Commits](https://github.com/awslabs/aws-sdk-rust/commits)

---
updated-dependencies:
- dependency-name: aws-sdk-sts
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

dependabot[bot] committed:
chore(deps): bump aws-sdk-sso from 1.50.0 to 1.56.0 in /datafusion-cli (#14259)

Bumps [aws-sdk-sso](https://github.com/awslabs/aws-sdk-rust) from 1.50.0 to 1.56.0.
- [Release notes](https://github.com/awslabs/aws-sdk-rust/releases)
- [Commits](https://github.com/awslabs/aws-sdk-rust/commits)

---
updated-dependencies:
- dependency-name: aws-sdk-sso
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

dependabot[bot] committed:
chore(deps): bump korandoru/hawkeye from 5 to 6 (#14354)

Bumps [korandoru/hawkeye](https://github.com/korandoru/hawkeye) from 5 to 6.
- [Release notes](https://github.com/korandoru/hawkeye/releases)
- [Changelog](https://github.com/korandoru/hawkeye/blob/main/CHANGELOG.md)
- [Commits](https://github.com/korandoru/hawkeye/compare/v5...v6)

---
updated-dependencies:
- dependency-name: korandoru/hawkeye
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

berkaysynnada committed:
remove one row sorts

ozankabak committed:
Improve comments

dependabot[bot] committed:
chore(deps): bump aws-sdk-ssooidc in /datafusion-cli (#14369)

Bumps [aws-sdk-ssooidc](https://github.com/awslabs/aws-sdk-rust) from 1.57.1 to 1.58.0.
- [Release notes](https://github.com/awslabs/aws-sdk-rust/releases)
- [Commits](https://github.com/awslabs/aws-sdk-rust/commits)

---
updated-dependencies:
- dependency-name: aws-sdk-ssooidc
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

alamb committed:
Minor: include the number of files run in sqllogictest display (#14359)

* Minor: include the number of files run in sqllogictest display

* fmt

* Update datafusion/sqllogictest/bin/sqllogictests.rs

ozankabak committed:
Use a short name for set monotonicity

buraksenn committed:
removed (#14370)

logan-keede committed:
move information_schema to datafusion-catalog (#14364)

* move information_schema to datafusion-catalog

* fix: formatting

* fix: doctests import

* Remove unecessary datafuson-catalog dependency

* remove some more unecessary dependencies

* Update datafusion-cli/Carglo.ock

* fix: doctest dependency

---------

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

alamb committed:
Unpin aws sdk dependencies (#14361)

findepi committed:
Core: Fix incorrect searched CASE optimization (#14349)

* Fix incorrect searched CASE optimization

There is an optimization for searched CASE where values are of boolean
type. It was converting the expression like

    CASE
        WHEN X THEN A
        WHEN Y THEN B
        ..
        [ ELSE D ]
    END

into

    (X AND A)
        OR (Y AND NOT X AND B)
        [ OR (NOT (X OR Y) AND D) ]

This had the following problems

- does not work for nullable conditions. If X is nullable, we cannot use
  NOT (X) to compliment it. We need to use `X IS DISTINCT FROM true`
- it does not work correctly when some conditions are nullable and other
  values are false. E.g. X=NULL, A=true, Y=NULL, B=true, D=false, the
  CASE should return false, but the boolean expression will simplify to
  `(NULL AND ..) OR (NULL AND ..) OR (false)` which is NULL, not false
  - thus we use `X` for truthness check of `X`, we need to test `X IS
    NOT DISTINCT FROM true`
- it did not work correctly when default D is missing, but conditions
  do not evaluate to NULL. CASE's result should be NULL but was false.

This commit fixes that optimization.

* Fix complexity comment

Rachelint committed:
Improve speed of  `median` by implementing special `GroupsAccumulator`  (#13681)

* draft of `MedianGroupAccumulator`.

* impl `state`.

* impl rest methods of `MedianGroupsAccumulator`.

* improve comments.

* use `MedianGroupsAccumulator`.

* remove unused import.

* add `group_median_table` to test group median.

* complete group median test cases in aggregate slt.

* fix type of state.

* Clippy

* Fmt

* add fuzzy tests for median.

* fix decimal.

* fix clippy.

* improve comments.

* add median cases with nulls.

* Update datafusion/functions-aggregate/src/median.rs

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

* use `OffsetBuffer::new_unchecked` in `convert_to_state`.

* add todo.

* remove assert and switch to i32 try from.

* return error when try from failed.

---------

Co-authored-by: Daniël Heres <danielheres@gmail.com>
Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

ozankabak committed:
Merge branch 'main' into feature/monotonic-sets

berkaysynnada commented:
https://github.com/apache/datafusion/pull/14271# merged, closing this one.
